### ISSUE TITLE
`_PickleCore._get_cache_by_key()` returns dict but is used as a dataclass

### ISSUE THREAD

**Issue by Leinadium** [python-cachier/cachier#246]
This library is required by [copernicusmarine](https://pypi.org/project/copernicusmarine/), and recently one of my data pipelines raised an exception with the following message:

    dict has no attribute "._completed"

Unfortunaly I couldn't get the full exception, so I had to trace all references to the `._completed` attribute, and found this python module.

While checking the code, I found that the `PickleCore`  `get_entry_by_key()` internally [calls](https://github.com/python-cachier/cachier/blob/master/src/cachier/cores/pickle.py#L159) `_get_cache_by_key()`, expecting a return type of  `CacheEntry`. But the `_get_cache_by_key()` method has a return type of `Dict[str, CacheEntry]]`

I would open a PR, but I couldn't understand what is the expected return type of `_get_cache_by_key()`.

This bug seems to be triggered whenever `cachier()` is called with the pickle implementation and with the `separate_files` set as True.

**Comment by shaypal5** [comment]
@Borda, this seems related to the new types you defined.

**Comment by Borda** [comment]
just checking since we have tests for this but all CI is 🟢 
so seems we have issue with testing, I see error treated as warning
https://github.com/python-cachier/cachier/actions/runs/11446891113/job/31847045677#step:5:219

**Comment by Borda** [comment]
ok, seems I have another example here: https://github.com/Lightning-AI/torchmetrics/actions/runs/11454171728/job/31867880744?pr=2797

**Comment by Leinadium** [comment]
After some examination, from my understanding:

[`PickleCore.get_entry_by_key`](https://github.com/python-cachier/cachier/blob/master/src/cachier/cores/pickle.py#L154) is supposed to return an entry (and its own key).

If `reload` is set to True, the method calls `_get_cache()`, which returns a cache (dict), and then calls `get` to get an entry.
But, if `separate_files` is set to True, then it just calls `get_cache_by_key`, which returns a full cache, not just a cache entry.

From my understanding, as both the cache and cache entry were dicts before, whichever function called `get_entry_by_key` would attempt to access some field thinking it was a cache entry, fail as it is a full cache, and handle it as a cache miss. This could be why the tests weren't triggered before (although some tests were failing as warnings #247)

**Comment by Borda** [comment]
> From my understanding, as both the cache and cache entry were dicts before, whichever function called `get_entry_by_key` would attempt to access some field thinking it was a cache entry, fail as it is a full cache, and handle it as a cache miss. This could be why the tests weren't triggering before (although some tests were failing as warnings #247)

hmm think that with the refactor to entry we sufficed some hidden issues 🐰

**Comment by veenstrajelmer** [comment]
I also encountered something similar here: https://github.com/mercator-ocean/copernicus-marine-toolbox/issues/183
This error only occurs with cachier 3.1.1, not with 3.0.1. So could it be that there is a bug introduced in that version?

**Comment by Borda** [comment]
> found out that during tests, `_PickleCore._get_cache_by_key` does return a `CacheEntry`, and not a cache dict! So maybe this is the intended behaviour, and the type hint is just wrong

do you have minimal example to reproduce which we can add to tests?

**Comment by Borda** [comment]
ok thinking this could be a transition issue as your past cache is `dict` bt the new/expected is `dataclass`

**Comment by Leinadium** [comment]
> @Leinadiumdo you have minimal example to reproduce which we can add to tests?

The tests are all fine. The tests expect that `_get_cache_by_key` returns a `CacheEntry`, and they succeded just fine. I modified the code to return a cache dict, and they started to fail. So the implementation is correct, just the type hint is wrong.



> right, this could be a transition issue as your past cache is `dict` but the new/expected is `dataclass`

Other developers noticed the same problem [here](https://github.com/mercator-ocean/copernicus-marine-toolbox/issues/183), and yes, caches are now incompatible between versions.

**Comment by Borda** [comment]
> Other developers noticed the same problem [here](https://github.com/mercator-ocean/copernicus-marine-toolbox/issues/183), and yes, caches are now incompatible between versions.

going to fix it now, just thinking about how to test it :/

**Comment by Leinadium** [comment]
> I see so we shall consider adding some typing check like `mypy`

That might be a good idea! I opened the project in my local IDE, and some static and type checks starting popping (e. g. multiple "possible None being used" values). It is a good idea!

> going to fix it now, just thinking about how to test it :/

The pickle object is being returned without any checks. You may add some checks to see if the object is a `CacheEntry`, and then for testing, you can create a bad/corrupt pickle file, and check for empty caches (and not failures).

**Comment by Borda** [comment]
> The pickle object is being returned without any checks. You may add some checks to see if the object is a `CacheEntry`

that is fine and simple

> and then for testing, you can create a bad/corrupt pickle file, and check for empty caches (and not failures).

I was rather thinking about running examples with several versions installed

**Comment by Borda** [comment]
@Leinadium could you test it also on your side with the open PR:
```
cachier @ https://github.com/python-cachier/cachier/archive/refs/heads/fix/legacy.zip
```

**Comment by shaypal5** [comment]
Fixed in release `v3.1.2`. Please reopen if that is not the case.
